### PR TITLE
feat: textArea -  add word count ( same as antd mobile)  example: 20/500

### LINKED ISSUE #24940: Display character counter for input component optionally
- [x] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### What problem does this feature solve?
I think having a character counter when there is a limit on the input field would be very important for UX and useful for projects using ant. It's a very common practice and material design promotes it too. Of course you can use the suffix or addon props with a bit of custom logic to achieve this, but that would create some problems. It would not be possible for the TextArea component, it wouldn't have the option of showing the counter outside of the input, and most importantly I find it sub-optimal to have to create custom components that wrap ant inputs in the projects that need this and have to use those instead.
I'd be more than happy to do this and create a PR for it if you think it's a good idea.

### What does the proposed API look like?
Perhaps we can have a showCounter boolean prop to the input which would show the number of characters entered if there is no maxLength prop, and show the ratio of characters used and the total character limit if a maxLength was specified. I attached a design for how the counter might look like.

Screen Shot 2020-06-12 at 6 20 04 AM

### PR DESCRIPTION
**Pull request by komagic**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

在业务开发中，经常需要使用到textArea 的计数功能（如：20/100），之前发现antd-mobile有count字数功能，而antd-design没有，就考虑加上一个。
### 💡 Background and solution

截屏2020-07-19 上午11 54 18

close #24940

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

-----
[View rendered components/input/index.en-US.md](https://github.com/komagic/ant-design/blob/feature/components/input/index.en-US.md)
[View rendered components/input/index.zh-CN.md](https://github.com/komagic/ant-design/blob/feature/components/input/index.zh-CN.md)

### CODE REVIEW

**Review by fireairforce — COMMENTED** [review]
加一下 test case 吧~

**Comment by fireairforce** [comment]
lint也挂了，修改一下。

File: components/input/TextArea.tsx [hunk 1]
```diff
@@ -71,23 +77,34 @@ class TextArea extends React.Component<TextAreaProps, TextAreaState> {
     resolveOnChange(this.resizableTextArea.textArea, e, this.props.onChange);
   };
 
-  renderTextArea = (prefixCls: string, bordered: boolean) => {
+  renderTextArea = (prefixCls: string, bordered: boolean,hasCount:boolean,value:string,maxLength:number) => {
+    let characterLength = countSymbols(value);
```
Comment (07akioni) [inline components/input/TextArea.tsx:81]: make it const

File: components/input/index.en-US.md [hunk 2]
```diff
@@ -49,6 +49,8 @@ The rest of the props of Input are exactly the same as the original [input](http
 | allowClear | If allow to remove input content with clear icon | boolean | false |  |
 | onResize | The callback function that is triggered when resize | function({ width, height }) | - |  |
 | bordered | Whether has border style | boolean | true | 4.5.0 |
+| hasCount | Whether has word count | boolean | false |  |
```
Comment (07akioni) [inline components/input/index.en-US.md:52]: please add version 4.5.0

File: components/input/index.zh-CN.md [hunk 3]
```diff
@@ -50,6 +50,7 @@ Input 的其他属性和 React 自带的 [input](https://facebook.github.io/reac
 | allowClear | 可以点击清除图标删除内容 | boolean | false |  |
 | onResize | resize 回调 | function({ width, height }) | - |  |
 | bordered | 是否有边框 | boolean | true | 4.5.0 |
+| hasCount | 是否显示字数统计 | boolean | false |  |
```
Comment (07akioni) [inline components/input/index.zh-CN.md:53]: add version

**Review by 07akioni — COMMENTED** [review]
Fix lint & add test cases

**Comment by komagic** [comment]
1. added two test count words and without hasCount 
2. fixed lint

File: components/input/TextArea.tsx [hunk 4]
```diff
@@ -71,23 +77,45 @@ class TextArea extends React.Component<TextAreaProps, TextAreaState> {
     resolveOnChange(this.resizableTextArea.textArea, e, this.props.onChange);
   };
 
-  renderTextArea = (prefixCls: string, bordered: boolean) => {
+  renderTextArea = (
+    prefixCls: string,
+    bordered: boolean,
+    hasCount: boolean,
+    value: string,
+    maxLength: number|undefined,
```
Comment (xrkffgg) [inline components/input/TextArea.tsx:85]:
```suggestion
    maxLength: number | undefined,
```

File: components/input/index.en-US.md [hunk 5]
```diff
@@ -49,6 +49,8 @@ The rest of the props of Input are exactly the same as the original [input](http
 | allowClear | If allow to remove input content with clear icon | boolean | false |  |
 | onResize | The callback function that is triggered when resize | function({ width, height }) | - |  |
 | bordered | Whether has border style | boolean | true | 4.5.0 |
+| hasCount | Whether has word count | boolean | false | 4.5.0 |
+
```
Comment (xrkffgg) [inline components/input/index.en-US.md:53]:
```suggestion

```

File: components/input/TextArea.tsx (hunk 4 above)
Comment (komagic) [reply components/input/TextArea.tsx]: thanks

File: components/input/TextArea.tsx [hunk 6]
```diff
@@ -9,15 +9,21 @@ import { fixControlledValue, resolveOnChange } from './Input';
...
   bordered?: boolean;
+  hasCount?: boolean;
 }
 
 export interface TextAreaState {
   value: any;
 }
 
+function countSymbols(text = ''): number {
+  if (text == null || text == undefined) return 0;
```
Comment (07akioni) [inline components/input/TextArea.tsx:20]: `if (!text) return 0;` is enough, there is no way for `text === undefined`

File: components/input/TextArea.tsx [hunk 7]
```diff
@@ -71,23 +77,45 @@ class TextArea extends React.Component<TextAreaProps, TextAreaState> {
...
   renderComponent = ({ getPrefixCls, direction }: ConfigConsumerProps) => {
-    const { value } = this.state;
-    const { prefixCls: customizePrefixCls, bordered = true } = this.props;
+    const { value = '' } = this.state;
+    const {
+      prefixCls: customizePrefixCls,
+      bordered = true,
+      hasCount = false,
+      maxLength = hasCount ? 500 : undefined,
+    } = this.props;
```
Comment (07akioni) [inline components/input/TextArea.tsx:118]: why `maxLength` is not documented?

File: components/input/TextArea.tsx (hunk 6 above)
Comment (komagic) [reply components/input/TextArea.tsx:20]: thanks,what should I do, only keep typeof text ===’null‘ ?

File: components/input/TextArea.tsx (hunk 7 above)
Comment (komagic) [reply components/input/TextArea.tsx:118]: I thought maxLength has been documented by Input 
截屏2020-07-22 上午10 41 31

File: components/input/TextArea.tsx [hunk 8]
```diff
@@ -9,15 +9,21 @@ import { fixControlledValue, resolveOnChange } from './Input';
 export interface TextAreaProps extends RcTextAreaProps {
   allowClear?: boolean;
   bordered?: boolean;
+  hasCount?: boolean;
```
Comment (shaodahong) [inline components/input/TextArea.tsx:12]: Is better naming `showCount`?

Comment (komagic) [reply components/input/TextArea.tsx:12]: The name is from antd-mobile, i prefer yours or enableCount

Comment (shaodahong) [reply components/input/TextArea.tsx:12]: https://github.com/ant-design/ant-design/wiki/API-Naming-rules#props

@07akioni What do you think?

File: components/input/TextArea.tsx (hunk 6 above)
Comment (07akioni) [reply components/input/TextArea.tsx:20]: that is ok

File: components/input/TextArea.tsx (hunk 8 above)
Comment (07akioni) [reply components/input/TextArea.tsx:12]: `showCount`

File: components/input/TextArea.tsx (hunk 7 above)
Comment (07akioni) [reply components/input/TextArea.tsx:118]: You are adding feature to `Input.TextArea`, not `Input` itself. So adding it to API docs is in need.

File: components/input/index.zh-CN.md [hunk 9]
```diff
@@ -50,7 +50,8 @@ Input 的其他属性和 React 自带的 [input](https://facebook.github.io/reac
 | allowClear | 可以点击清除图标删除内容 | boolean | false |  |
 | onResize | resize 回调 | function({ width, height }) | - |  |
 | bordered | 是否有边框 | boolean | true | 4.5.0 |
-| hasCount | 是否显示字数统计 | boolean | false | 4.5.0 |
+| showCount | 是否显示字数统计 | boolean | false | 4.5.0 |
+| maxLength | 最大长度 | number | - | 4.5.0 |
```
Comment (07akioni) [inline components/input/index.zh-CN.md:54]: how about default 500. Works when only `showCount` is `true`? The behavior doesn't follow docs now.

Comment (komagic) [reply components/input/index.zh-CN.md]: 大佬，1、我就是想在有showCount的时候，maxlength才显示，该如何优雅的code出来？ 2、测试总是报错误，有啥办法修改？

Comment (07akioni) [reply components/input/index.zh-CN.md]: 我觉得是这样，默认 maxLength 为空，showCount 只展示字数，然后如果加上一起展示，而且对非 showCount 状况也生效。

Comment (07akioni) [reply components/input/index.zh-CN.md]: 现在这样行为不一致

Comment (07akioni) [reply components/input/index.zh-CN.md]: 也不清晰

Comment (07akioni) [reply components/input/index.zh-CN.md]: 测试的话本地测一下或者点 details，看看自己哪个用例挂了，改掉

File: components/comment/__tests__/__snapshots__/demo.test.js.snap [hunk 10]
```diff
@@ -156,10 +156,15 @@ exports[`renders ./components/comment/demo/editor.md correctly 1`] = `
               <div
                 class="ant-form-item-control-input-content"
               >
-                <textarea
-                  class="ant-input"
-                  rows="4"
-                />
+                <div
+                  class="ant-input-control"
+                  value=""
```
Comment (xiaohuoni) [inline components/comment/__tests__/__snapshots__/demo.test.js.snap:161]: why value=""?

**Comment by 07akioni** [comment]
CI Failed

**Comment by afc163** [comment]
Closing for no response, feel free to recreate another PR.

**Comment by mxye** [comment]
So, do I understand it correctly that "showCount" will not be added as a feature to the generic Input, and if I would still like "showCount", a workaround would be needed, such as using "TextArea" but limiting the number of Rows to 1 to mimic the generic Input? -- The TextArea doesn't exactly look the same as generic Input, and in some cases generic Input would look better -- if only it has showCount...

**Pull request closed without merging**